File: app/src/main/java/org/rm3l/router_companion/service/BackgroundService.java
package org.rm3l.router_companion.service;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.util.Log;

import com.crashlytics.android.Crashlytics;

import org.rm3l.router_companion.mgmt.BootReceiver;
import org.rm3l.router_companion.mgmt.RouterManagementActivity;
import org.rm3l.router_companion.mgmt.dao.DDWRTCompanionDAO;
import org.rm3l.router_companion.resources.conn.Router;
import org.rm3l.router_companion.service.tasks.AbstractBackgroundServiceTask;
import org.rm3l.router_companion.service.tasks.ConnectedHostsServiceTask;
import org.rm3l.router_companion.service.tasks.PublicIPChangesServiceTask;
import org.rm3l.router_companion.service.tasks.RouterInfoForFeedbackServiceTask;
import org.rm3l.router_companion.service.tasks.RouterModelUpdaterServiceTask;
import org.rm3l.router_companion.service.tasks.RouterWebInterfaceParametersUpdaterServiceTask;
import org.rm3l.router_companion.utils.DDWRTCompanionConstants;
import org.rm3l.router_companion.utils.ReportingUtils;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created by rm3l on 05/07/15.
 */
public class BackgroundService extends IntentService {

    private static final String TAG = BackgroundService.class.getSimpleName();

//    private PowerManager.WakeLock mWakeLock;

    public BackgroundService() {
        super(BackgroundService.class.getSimpleName());
    }

    /**
     * Creates an IntentService.  Invoked by your subclass's constructor.
     *
     * @param name Used to name the worker thread, important only for debugging.
     */
    public BackgroundService(String name) {
        super(name);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        try {
        // check the global background data setting
            final SharedPreferences sharedPreferences =
                    getSharedPreferences(DDWRTCompanionConstants.DEFAULT_SHARED_PREFERENCES_KEY, Context.MODE_PRIVATE);
            sharedPreferences.edit()
                    .putLong(DDWRTCompanionConstants.BG_SERVICE_LAST_HANDLE, System.currentTimeMillis())
                    .apply();

            final ConnectivityManager cm = (ConnectivityManager) getSystemService(CONNECTIVITY_SERVICE);
            if (intent == null || cm == null || cm.getActiveNetworkInfo() == null) {
            /*
             * If getActiveNetworkInfo() is null, you do not have a network connection,
             * either because the device does not have a network connection,
             * or because user settings (e.g., bandwidth caps) prevent your app
             * from having a network connection.
             */
//                stopSelf();
                return;
            }

            //Do the actual work - no need to do so in a separate thread
            final List<AbstractBackgroundServiceTask> tasks = new ArrayList<>();
            tasks.add(new RouterModelUpdaterServiceTask(BackgroundService.this));
            tasks.add(new RouterInfoForFeedbackServiceTask(BackgroundService.this));
            tasks.add(new RouterWebInterfaceParametersUpdaterServiceTask(BackgroundService.this));
            //According to user preference
            final Set<String> notificationsChoiceSet =
                    sharedPreferences
                            .getStringSet(DDWRTCompanionConstants.NOTIFICATIONS_CHOICE_PREF, new HashSet<String>());
            Crashlytics.log(Log.DEBUG,  TAG, "notificationsChoiceSet: " + notificationsChoiceSet);
            if (notificationsChoiceSet.contains(ConnectedHostsServiceTask.class.getSimpleName())) {
                tasks.add(new ConnectedHostsServiceTask(BackgroundService.this));
            }
            if (notificationsChoiceSet.contains(PublicIPChangesServiceTask.class.getSimpleName())) {
                tasks.add(new PublicIPChangesServiceTask(BackgroundService.this));
            }
            //Add any other tasks over here

            final DDWRTCompanionDAO dao = RouterManagementActivity.getDao(BackgroundService.this);
            final List<Router> allRouters = dao.getAllRouters();
            if (allRouters == null) {
                return;
            }
            for (final Router router : allRouters) {
                if (router == null) {
                    continue;
                }

                //Execute tasks available for notifications
                for (final AbstractBackgroundServiceTask backgroundServiceTask : tasks) {
                    if (backgroundServiceTask == null) {
                        continue;
                    }
                    Crashlytics.log(Log.DEBUG,  TAG, ">>> Running task: " +
                            backgroundServiceTask.getClass() + " on router " + router);
                    try {
                        backgroundServiceTask.runBackgroundServiceTask(router);
                    } catch (Exception e) {
                        e.printStackTrace();
                        ReportingUtils.reportException(this, e);
                        //No worries
                    }
                }
            }

        } finally {
            try {
                BackgroundService.this.stopSelf();
            } finally {
                // Release the wake lock provided by the WakefulBroadcastReceiver.
                if (intent != null) {
                    BootReceiver.completeWakefulIntent(intent);
                }
            }
        }
    }
}

File: app/src/main/java/org/rm3l/router_companion/service/BackgroundService.kt
package org.rm3l.router_companion.service

import android.content.Context
import android.net.ConnectivityManager
import com.evernote.android.job.DailyJob
import com.evernote.android.job.Job
import com.evernote.android.job.JobManager
import com.evernote.android.job.JobRequest
import com.google.firebase.crashlytics.FirebaseCrashlytics
import org.rm3l.ddwrt.BuildConfig
import org.rm3l.router_companion.RouterCompanionAppConstants
import org.rm3l.router_companion.job.RouterCompanionJob
import org.rm3l.router_companion.mgmt.RouterManagementActivity
import org.rm3l.router_companion.service.tasks.AbstractBackgroundServiceTask
import org.rm3l.router_companion.service.tasks.ConnectedHostsServiceTask
import org.rm3l.router_companion.service.tasks.PublicIPChangesServiceTask
import org.rm3l.router_companion.service.tasks.RouterInfoForFeedbackServiceTask
import org.rm3l.router_companion.service.tasks.RouterModelUpdaterServiceTask
import org.rm3l.router_companion.service.tasks.RouterWebInterfaceParametersUpdaterServiceTask
import org.rm3l.router_companion.utils.Utils
import java.util.concurrent.TimeUnit

/**
 * Created by rm3l on 05/07/15.
 */
class BackgroundService : DailyJob(), RouterCompanionJob {

    companion object {
        @JvmField
        val TAG = BackgroundService::class.java.simpleName

        @JvmStatic
        fun schedule() {
            if (!JobManager.instance().getAllJobRequestsForTag(TAG).isEmpty()) {
                // job already scheduled, nothing to do
                FirebaseCrashlytics.getInstance().log("job $TAG already scheduled => nothing to do!")
                return
            }
            val builder = JobRequest.Builder(TAG)
                .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                .setRequiresBatteryNotLow(true)
            // run job between 6am and 5am
            DailyJob.schedule(
                builder,
                TimeUnit.HOURS.toMillis(6),
                TimeUnit.HOURS.toMillis(5)
            )
        }

        @JvmStatic
        fun handleJob(context: Context, params: Params?) {
            val dao = RouterManagementActivity.getDao(context)
            val globalPreferences = context.getSharedPreferences(
                RouterCompanionAppConstants.DEFAULT_SHARED_PREFERENCES_KEY,
                Context.MODE_PRIVATE
            )

            // check the global background data setting
            globalPreferences?.edit()
                ?.putLong(RouterCompanionAppConstants.BG_SERVICE_LAST_HANDLE, System.currentTimeMillis())
                ?.apply()
            Utils.requestBackup(context)

            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            if (cm.activeNetworkInfo != null) {
                /*
                 * If getActiveNetworkInfo() is null, you do not have a network connection,
                 * either because the device does not have a network connection,
                 * or because user settings (e.g., bandwidth caps) prevent your app
                 * from having a network connection.
                 */
                // Do the actual work - no need to do so in a separate thread
                val tasks: MutableList<AbstractBackgroundServiceTask> = arrayListOf()
                tasks.add(RouterModelUpdaterServiceTask(context))
                tasks.add(RouterInfoForFeedbackServiceTask(context))
                tasks.add(RouterWebInterfaceParametersUpdaterServiceTask(context))
                if (BuildConfig.DONATIONS) {
                    FirebaseCrashlytics.getInstance().log(
                        "ConnectedHostsServiceTask and PublicIPChangesServiceTask background notifications" +
                            " are *Premium* features!"
                    )
                } else {
                    // According to user preference
                    val notificationsChoiceSet = globalPreferences?.getStringSet(
                        RouterCompanionAppConstants.NOTIFICATIONS_CHOICE_PREF,
                        emptySet()
                    ) ?: emptySet()
                    FirebaseCrashlytics.getInstance().log("notificationsChoiceSet: " + notificationsChoiceSet)
                    if (notificationsChoiceSet.contains(ConnectedHostsServiceTask::class.java.simpleName)) {
                        tasks.add(ConnectedHostsServiceTask(context))
                    }
                    if (notificationsChoiceSet.contains(PublicIPChangesServiceTask::class.java.simpleName)) {
                        tasks.add(PublicIPChangesServiceTask(context))
                    }
                }
                // Add any other tasks over here

                val allRouters = dao.allRouters
                for (router in allRouters) {
                    // Execute tasks available for notifications
                    for (backgroundServiceTask in tasks) {
                        FirebaseCrashlytics.getInstance().log(
                            ">>> Running task: ${backgroundServiceTask.javaClass} on router $router"
                        )
                        try {
                            backgroundServiceTask.runBackgroundServiceTask(router)
                        } catch (e: Exception) {
                            FirebaseCrashlytics.getInstance().recordException(e)
                        }
                    }
                }
            }
        }
    }

    override fun isOneShotJob() = false

    override fun onRunDailyJob(params: Params): DailyJobResult {
        try {
            if (context.getSharedPreferences(
                    RouterCompanionAppConstants.DEFAULT_SHARED_PREFERENCES_KEY, Context.MODE_PRIVATE
                )
                .getBoolean(RouterCompanionAppConstants.NOTIFICATIONS_BG_SERVICE_ENABLE, false)
            ) {
                handleJob(context, params)
            } else {
                FirebaseCrashlytics.getInstance().log("Background Service disabled (user choice)")
            }
        } catch (e: Exception) {
            // No worries
            FirebaseCrashlytics.getInstance().recordException(e)
        }
        return DailyJobResult.SUCCESS
    }
}

class BackgroundServiceOneShotJob : Job(), RouterCompanionJob {

    companion object {
        val TAG = BackgroundServiceOneShotJob::class.java.simpleName
    }

    override fun onRunJob(params: Params): Result {
        return try {
            BackgroundService.handleJob(context, params)
            Result.SUCCESS
        } catch (e: Exception) {
            Result.FAILURE
        }
    }
}
